File: internal/httpx/reply_feed_context.go
package httpx

import (
	"html"
	"html/template"
	"strconv"
	"strings"

	"ptxt-nstr/internal/nostrx"
	"ptxt-nstr/internal/thread"
)

// isFeedThreadReply is true when the text note replies to another note in the
// thread (NIP-10), using the same RootID/ParentID rules as the thread builder.
func isFeedThreadReply(ev nostrx.Event) bool {
	if ev.Kind != nostrx.KindTextNote {
		return false
	}
	root := thread.RootID(ev)
	parent := thread.ParentID(root, ev)
	if parent == "" {
		return false
	}
	return !strings.EqualFold(strings.TrimSpace(parent), strings.TrimSpace(ev.ID))
}

// replyContextVisible is true when the feed should show the "Replying to …" row.
func replyContextVisible(ev nostrx.Event) bool {
	return isFeedThreadReply(ev) && !isQuotePost(ev)
}

// replyContextTargets returns ordered unique p-tag pubkeys, excluding the author.
func replyContextTargets(ev nostrx.Event) []string {
	author := strings.TrimSpace(ev.PubKey)
	seen := make(map[string]bool)
	var out []string
	for _, tag := range ev.Tags {
		if len(tag) < 2 || tag[0] != "p" {
			continue
		}
		pk, err := nostrx.NormalizePubKey(tag[1])
		if err != nil {
			continue
		}
		if seen[pk] {
			continue
		}
		if author != "" && strings.EqualFold(pk, author) {
			continue
		}
		seen[pk] = true
		out = append(out, pk)
	}
	return out
}

func replyMentionLink(profiles map[string]nostrx.Profile, pubkey string) string {
	pk := thread.NormalizeHexEventID(strings.TrimSpace(pubkey))
	if len(pk) != 64 {
		return ""
	}
	label := authorLabel(profiles, pk)
	if label == "" {
		label = short(pk)
	}
	escaped := html.EscapeString("@" + label)
	href := "/u/" + html.EscapeString(pk)
	return `<a href="` + href + `" data-relay-aware>` + escaped + `</a>`
}

// replyContextHTML builds safe HTML for the reply context row (no outer wrapper).
func replyContextHTML(profiles map[string]nostrx.Profile, ev nostrx.Event) template.HTML {
	if !replyContextVisible(ev) {
		return ""
	}
	root := thread.RootID(ev)
	parent := thread.ParentID(root, ev)
	parent = thread.NormalizeHexEventID(strings.TrimSpace(parent))

	targets := replyContextTargets(ev)
	var b strings.Builder
	b.WriteString(`<span class="note-feed-context-lead">Replying to </span>`)

	if len(targets) == 0 {
		if len(parent) != 64 {
			return template.HTML(b.String())
		}
		b.WriteString(`<a href="/thread/`)
		b.WriteString(html.EscapeString(parent))
		b.WriteString(`" data-relay-aware>thread</a>`)
		return template.HTML(b.String())
	}

	show := targets
	rest := 0
	if len(show) > 2 {
		rest = len(show) - 2
		show = show[:2]
	}
	for i, pk := range show {
		if i > 0 {
			b.WriteString(` `)
		}
		if link := replyMentionLink(profiles, pk); link != "" {
			b.WriteString(link)
		}
	}
	if rest > 0 {
		b.WriteString(` <span class="note-feed-context-tail">and `)
		b.WriteString(strconv.Itoa(rest))
		b.WriteString(` other`)
		if rest != 1 {
			b.WriteString(`s`)
		}
		b.WriteString(`</span>`)
	}
	return template.HTML(b.String())
}

// repostContextHTML builds safe HTML for the repost banner (no outer wrapper).
func repostContextHTML(profiles map[string]nostrx.Profile, ev nostrx.Event) template.HTML {
	if ev.Kind != nostrx.KindRepost {
		return ""
	}
	pk := thread.NormalizeHexEventID(strings.TrimSpace(ev.PubKey))
	name := html.EscapeString(displayName(profiles, pk))
	var b strings.Builder
	b.WriteString(`<span class="note-feed-context-repost-inner">`)
	b.WriteString(name)
	b.WriteString(` reposted</span>`)
	return template.HTML(b.String())
}

File: web/static/js/reply-feed-context.js
import { KIND_NOTE } from "./nostr-kinds.js";
import { isQuotePost } from "./note-references.js";
import { displayName } from "./profile-parse.js";
import { parentID } from "./thread-tags.js";
import { canonicalHex64, normalizePubkey, profilePath } from "./relay-utils.js";

function shortPubkey(pubkey) {
  if (!pubkey) return "";
  if (pubkey.length <= 12) return pubkey;
  return `${pubkey.slice(0, 8)}…${pubkey.slice(-4)}`;
}

function escapeHtml(value) {
  return String(value || "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

/** Mirrors thread.RootID (kind-1 only; empty when no e-tags). */
export function feedThreadRootID(event) {
  if (Number(event?.kind) !== KIND_NOTE) return "";
  for (const tag of event?.tags || []) {
    if (!Array.isArray(tag) || tag.length < 2 || tag[0] !== "e") continue;
    const marker = tag.length >= 4 ? String(tag[3] || "") : "";
    if (marker.toLowerCase() === "root") return canonicalHex64(tag[1]);
  }
  for (const tag of event?.tags || []) {
    if (!Array.isArray(tag) || tag.length < 2 || tag[0] !== "e") continue;
    return canonicalHex64(tag[1]);
  }
  return "";
}

/** Mirrors httpx.isFeedThreadReply. */
export function isFeedThreadReply(event) {
  if (Number(event?.kind) !== KIND_NOTE) return false;
  const root = feedThreadRootID(event);
  const parent = canonicalHex64(parentID(root, event));
  if (!parent) return false;
  return parent !== canonicalHex64(event?.id);
}

/** Mirrors httpx.replyContextVisible. */
export function replyContextVisible(event) {
  return isFeedThreadReply(event) && !isQuotePost(event);
}

/** Mirrors httpx.replyContextTargets. */
export function replyContextTargets(event) {
  const author = normalizePubkey(event?.pubkey);
  const seen = new Set();
  const out = [];
  for (const tag of event?.tags || []) {
    if (!Array.isArray(tag) || tag.length < 2 || tag[0] !== "p") continue;
    const pk = normalizePubkey(tag[1]);
    if (!pk || seen.has(pk)) continue;
    if (author && pk === author) continue;
    seen.add(pk);
    out.push(pk);
  }
  return out;
}

function profileLabel(profilesByPubkey, pubkey) {
  const pk = normalizePubkey(pubkey);
  const profile = profilesByPubkey?.[pk];
  if (profile) return displayName(profile);
  return shortPubkey(pk);
}

function replyMentionLinkHTML(profilesByPubkey, pubkey) {
  const pk = normalizePubkey(pubkey);
  if (pk.length !== 64) return "";
  const label = profileLabel(profilesByPubkey, pk);
  return `<a href="${escapeHtml(profilePath(pk))}" data-relay-aware>@${escapeHtml(label)}</a>`;
}

/** Safe HTML for the reply context row (mirrors httpx.replyContextHTML). */
export function replyContextHTML(event, profilesByPubkey = {}) {
  if (!replyContextVisible(event)) return "";

  const root = feedThreadRootID(event);
  const parent = canonicalHex64(parentID(root, event));
  const targets = replyContextTargets(event);
  let html = `<span class="note-feed-context-lead">Replying to </span>`;

  if (!targets.length) {
    if (parent.length !== 64) return html;
    return `${html}<a href="/thread/${escapeHtml(parent)}" data-relay-aware>thread</a>`;
  }

  const show = targets.length > 2 ? targets.slice(0, 2) : targets;
  const rest = targets.length > 2 ? targets.length - 2 : 0;
  html += show
    .map((pk) => replyMentionLinkHTML(profilesByPubkey, pk))
    .filter(Boolean)
    .join(" ");
  if (rest > 0) {
    html += ` <span class="note-feed-context-tail">and ${rest} other${rest === 1 ? "" : "s"}</span>`;
  }
  return html;
}

/** Safe HTML for the repost banner (mirrors httpx.repostContextHTML). */
export function repostContextHTML(event, profilesByPubkey = {}) {
  if (Number(event?.kind) !== 6) return "";
  const pk = normalizePubkey(event?.pubkey);
  const name = escapeHtml(profileLabel(profilesByPubkey, pk));
  return `<span class="note-feed-context-repost-inner">${name} reposted</span>`;
}

/** Adds feed reply/repost context elements expected by ascii.js (server partial parity). */
export function appendFeedContextElements(container, event, profilesByPubkey = {}) {
  if (!container || !event) return;

  const repostHTML = repostContextHTML(event, profilesByPubkey);
  if (repostHTML) {
    const banner = document.createElement("p");
    banner.className = "note-feed-context note-feed-context--repost";
    banner.innerHTML = repostHTML;
    container.append(banner);
  }

  const replyHTML = replyContextHTML(event, profilesByPubkey);
  if (replyHTML) {
    const tmpl = document.createElement("template");
    tmpl.className = "note-reply-context-tmpl";
    tmpl.innerHTML = replyHTML;
    container.append(tmpl);
  }
}
